### PR TITLE
React Native v.71 Compliance

### PR DESCRIPTION
**Pull request by asafkorem**
This PR resolves https://github.com/wix/Detox/issues/4066, https://github.com/wix/Detox/issues/4067, https://github.com/wix/Detox/issues/4068, https://github.com/wix/Detox/issues/4069 and https://github.com/wix/Detox/issues/4070.

### CODE REVIEW

File: detox/test/android/build.gradle [hunk 1]
```diff
@@ -38,23 +43,25 @@ buildscript {
...
+                // We don't want to fetch react-native from Maven Central as there are
+                // older versions over there.
+                content {
+                    excludeGroup "com.facebook.react"
+                }
             }
+            google()
+            mavenLocal()
         }
-        google()
```
Comment (d4vidi) [inline detox/test/android/build.gradle:56]: Makes sense, although it might be a good idea to keep `google()` and `mavenLocal()` even in RN ≥71

File: docs/introduction/partials/_getting-started-rn.md [hunk 2]
```diff
@@ -1,16 +1,15 @@
...
+
 :::info Note
 
-Mind that Detox works, at least officially, with the following React Native versions:
+While Detox should work out of the box with almost any React Native version from the past few minor releases, official support is provided for React Native versions `0.70.x` and `0.71.x` without React Native's ["New Architecture"](https://reactnative.dev/docs/the-new-architecture/landing-page).
 
-| iOS             | Android                                                                                                                |
-| --------------- | ---------------------------------------------------------------------------------------------------------------------- |
-| 0.68.x - 0.70.7 | 0.68.x - 0.70.7 -<br/>Visibility edge-case: see this [RN issue](https://github.com/facebook/react-native/issues/23870) |
+Official support means that we test Detox against these versions and potentially fix issues that are specific to them. This does not mean that Detox will not work with other versions, but rather that we do not test Detox against them and do not fix issues that are specific to them.
```
Comment (noomorph) [inline docs/introduction/partials/_getting-started-rn.md:9]: Wow wow wow, too long. I got confused while reading. Try to make it simpler somehow and x2 or x3 shorter. @asafkorem

Comment (asafkorem) [reply docs/introduction/partials/_getting-started-rn.md:9]: lol, OK, I can see that. I'll do it :+1:

Comment (asafkorem) [reply docs/introduction/partials/_getting-started-rn.md:9]: 2x shorter:
> Official support means we actively test Detox on certain versions and fix their unique issues. Detox might work with other versions, but we don't test or fix issues specific to them.

3x shorter (less information):
> Detox might work with other versions, but we don't test or fix issues specific to them.

File: README.md [hunk 3]
```diff
@@ -45,39 +45,26 @@ High velocity native mobile development requires us to adopt continuous integrat
...
 - **Made For CI:** Execute your E2E tests on CI platforms like Travis CI, Circle CI or Jenkins without grief.
 - **Runs on Devices:** Gain confidence to ship by testing your app on a device/simulator just like a real user (not yet supported on iOS).
 - **Test Runner Agnostic:** Detox provides a set of APIs to use with any test runner without it. It comes with [Jest](https://jestjs.io) integration out of the box.
 
-## Supported Versions
+## Supported React Native Versions
 
-### Environment
+Detox was built from the ground up to support React Native projects.
+While Detox should work out of the box with almost any React Native version from the past few minor releases, official support is provided for React Native versions `0.70.x` and `0.71.x` without React Native's ["New Architecture"](https://reactnative.dev/docs/the-new-architecture/landing-page).
```
Comment (asafkorem) [inline README.md:58]: @d4vidi regarding your comment - What do you think about:
> We officially support Detox for React Native versions **0.70.x** and **0.71.x**, without the ["New Architecture"](https://reactnative.dev/docs/the-new-architecture/landing-page). 

TBH I'm not a huge fan of the table since there is really no difference between iOS and Android in terms of our official support. If you want some border that will wrap the version numbers I can make a 1x1 table 😄

File: examples/demo-react-native/android/app/build.gradle [hunk 4]
```diff
@@ -1,22 +1,20 @@
...
-]
-
-apply from: '../../node_modules/react-native/react.gradle'
+apply plugin: 'com.facebook.react'
 
 android {
-    compileSdkVersion rootProject.ext.compileSdkVersion
+    namespace 'com.example'
+
     buildToolsVersion rootProject.ext.buildToolsVersion
```
Comment (d4vidi) [inline examples/demo-react-native/android/app/build.gradle:7]: todo (note to self)

File: README.md [hunk 5]
```diff
@@ -45,39 +45,26 @@ High velocity native mobile development requires us to adopt continuous integrat
...
-**OS:**
+Newer versions, as well as React Native's "New Architecture", may work with Detox, but have not been tested out yet by the Detox team.
 
-- **macOS:** v10.15 (Catalina) or higher
-  - **Xcode**: v11.0 or higher
-  - **iOS Simulator Runtime**: iOS v13.0 or higher
-- **Ubuntu linux:** v20.04 or higher
-- **Windows 10:** Version 1803 or higher
+Although we do not officially support older React Native versions, we do our best to keep Detox compatible with them.
+Also, in case of a problem with a more recent React Native version or with React Native's "New Architecture", please [submit an issue](https://github.com/wix/Detox/issues/new/choose) and we will do our best to help you out and to address the issue once we will support these versions officially.
```
Comment (asafkorem) [inline README.md:63]:
```suggestion
Although we do not officially support older React Native versions, we do our best to keep Detox compatible with them.

Also, in case of a problem with an unsupported version of React Native, please [submit an issue](https://github.com/wix/Detox/issues/new/choose) or write us in our [Discord server](https://discord.gg/CkD5QKheF5) and we will do our best to help out.
```

File: detox/android/build.gradle [hunk 6]
```diff
@@ -19,11 +21,14 @@ buildscript {
     ext.detoxKotlinVersion = ext.kotlinVersion
 
     repositories {
-        mavenCentral()
         google()
+        mavenCentral()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:7.2.1'
+        if (!rnInfo.isRN71OrNewer) {
```
Comment (asafkorem) [inline detox/android/build.gradle:28]: consider replacing all those properties with `rnInfo.minorVersion < 71`

File: detox/android/build.gradle [hunk 7]
```diff
@@ -34,11 +39,16 @@ buildscript {
 
 allprojects {
     repositories {
-        mavenLocal()
-        mavenCentral()
         google()
-        maven {
-            url "$projectDir/../../node_modules/react-native/android"
+        mavenCentral()
+        mavenLocal()
```
Comment (asafkorem) [inline detox/android/build.gradle:44]: let's consider to remove this

File: detox/android/detox/build.gradle [hunk 8]
```diff
@@ -12,6 +13,13 @@ def _minSdkVersion = _ext.has('minSdkVersion') ? _ext.minSdkVersion : 21
 def _kotlinVersion = _ext.has('detoxKotlinVersion') ? _ext.detoxKotlinVersion : _kotlinMinVersion
 def _kotlinStdlib = _ext.has('detoxKotlinStdlib') ? _ext.detoxKotlinStdlib : 'kotlin-stdlib-jdk8'
 
+// RN native code comes from either maven-local (in which case, need the *exact* version),
```
Comment (asafkorem) [inline detox/android/detox/build.gradle:16]:
```suggestion
// RN native code comes from either maven-central (in which case, need the *exact* version),
```

File: detox/android/rninfo.gradle [hunk 9]
```diff
@@ -1,21 +1,28 @@
...
+def rnVersion = getRNVersion()
+def rnMajorVer = getMajorVersion(rnVersion)
+println "[$project] RNInfo: detected React Native version: $rnVersion (major=$rnMajorVer)"
 
 project.ext.rnInfo = [
+    version: rnVersion,
     majorVersion: rnMajorVer,
     isRN64OrHigher: rnMajorVer >= 64,
     isRN66OrHigher: rnMajorVer >= 66,
     isRN68OrHigher: rnMajorVer >= 68,
```
Comment (asafkorem) [inline detox/android/rninfo.gradle:12]:
```suggestion
```

Comment (asafkorem) [reply detox/android/rninfo.gradle:12]: might be problematic

File: detox/test/android/build.gradle [hunk 10]
```diff
@@ -1,35 +1,44 @@
...
         isOfficialDetoxApp = true
         kotlinVersion = '1.6.21'
-        buildToolsVersion = '31.0.0'
-        compileSdkVersion = 31
-        targetSdkVersion = 31
+        buildToolsVersion = '33.0.0'
+        compileSdkVersion = 33
+        targetSdkVersion = 33
         minSdkVersion = 21
         appCompatVersion = (rnInfo.isRN68OrHigher ? '1.4.2' : '1.3.0')
```
Comment (asafkorem) [inline detox/test/android/build.gradle:19]:
```suggestion
        appCompatVersion = '1.4.2'
```

File: detox/test/android/build.gradle [hunk 11]
```diff
@@ -40,20 +49,25 @@ buildscript {
...
+                url "$rootDir/../../node_modules/react-native/android"
+            }
+            google()
+            mavenCentral() {
+                if (!rnInfo.isRN71OrHigher) {
+                    content {
+                        excludeGroup 'com.facebook.react'
+                    }
+                }
             }
```
Comment (asafkorem) [inline detox/test/android/build.gradle:68]:
```suggestion
            mavenCentral() {
                content {
                    excludeGroup 'com.facebook.react'
                }
            }
```

File: detox/test/package.json [hunk 12]
```diff
@@ -25,7 +28,9 @@
     "build:ios": "detox build -c ios.sim.release",
     "build:ios-debug": "detox build -c ios.sim.debug",
     "build:android": "detox build -c android.emu.release",
-    "build:android-debug": "detox build -c android.emu.debug",
+    "build:android:debug": "detox build -c android.emu.debug",
+    "build:android-rnLegacy": "detox build -c android.emu.release.rnLegacy",
+    "build:android-rnLegacy:debug": "detox build -c android.emu.debug.rnLegacy",
```
[the remark this answers is not in the document]
Comment (asafkorem) [reply detox/test/package.json:33]: `cat node_modules/react-native/package.json | jq -r '.version' | cut -d. -f2`

File: docs/contributing.md [hunk 13]
```diff
@@ -144,12 +144,16 @@ launchctl setenv PATH $PATH
 
 ###### 3. Android Native Unit-Tests
 
-Under `detox/android`:
+Under `detox`:
```
Comment (asafkorem) [inline docs/contributing.md:147]:
```suggestion
Under `detox/`:
```

File: docs/introduction/partials/_getting-started-rn.md (hunk 2 above)
Comment (asafkorem) [inline docs/introduction/partials/_getting-started-rn.md:9]:
```suggestion
Newer versions, as well as React Native's "New Architecture", may work with Detox, but have not been tested out yet by the Detox team.
```

File: docs/introduction/partials/_getting-started-rn.md [hunk 14]
```diff
@@ -1,16 +1,15 @@
...
 
-Newer versions, as well as React Native's ["New Architecture"](https://reactnative.dev/docs/the-new-architecture/landing-page), are most likely supported, but have not been tested out yet.
-Please [submit an issue](https://github.com/wix/Detox/issues/new/choose) if you find specific problems with more recent React Native versions.
+Newer versions, as well as React Native's "New Architecture", may work with Detox, but have not been tested out yet by the Detox team. In case of a problem with a more recent React Native version or with React Native's "New Architecture", please [submit an issue](https://github.com/wix/Detox/issues/new/choose) and we will do our best to help you out and to address the issue once we will support these versions officially.
```
Comment (asafkorem) [inline docs/introduction/partials/_getting-started-rn.md:11]:
```suggestion
Although we do not officially support older React Native versions, we do our best to keep Detox compatible with them.
```

File: docs/introduction/partials/_getting-started-rn.md [hunk 15]
```diff
@@ -1,16 +1,15 @@
...
 
-For support for React Native 0.71.x, please track [this issue](https://github.com/wix/Detox/issues/3884).
+Although we do not officially support older React Native versions, we do our best to keep Detox compatible with them to the extent possible, as long as it does not require us to compromise on the quality of Detox's codebase.
```
Comment (asafkorem) [inline docs/introduction/partials/_getting-started-rn.md:13]:
```suggestion
Also, in case of a problem with an unsupported version of React Native, please [submit an issue](https://github.com/wix/Detox/issues/new/choose) or write us in our [Discord server](https://discord.gg/CkD5QKheF5) and we will do our best to help out.
```

File: examples/demo-react-native/android/build.gradle [hunk 16]
```diff
@@ -23,21 +16,18 @@ buildscript {
...
-        mavenCentral()
         google()
-        maven {
-            url "$rootDir/../node_modules/react-native/android"
-        }
+        mavenCentral()
         maven {
             url "$rootDir/../node_modules/detox/Detox-android"
         }
+        mavenLocal()
```
Comment (asafkorem) [inline examples/demo-react-native/android/build.gradle:31]:
```suggestion
```

File: examples/demo-react-native/ios/Podfile [hunk 17]
```diff
@@ -13,15 +14,30 @@ target 'example' do
...
+    if ENV["REACT_NATIVE_VERSION"] && ENV["REACT_NATIVE_VERSION"].match(/0.6[6,7,8,9].*/)
+      react_native_post_install(installer)
+    else
+      react_native_post_install(
+        installer,
+        # Set `mac_catalyst_enabled` to `true` in order to apply patches
+        # necessary for Mac Catalyst builds
+        :mac_catalyst_enabled => false
+      )
+    end
```
Comment (asafkorem) [inline examples/demo-react-native/ios/Podfile:28]:
```suggestion
     react_native_post_install(
       installer,
       # Set `mac_catalyst_enabled` to `true` in order to apply patches
       # necessary for Mac Catalyst builds
       :mac_catalyst_enabled => false
     )
```

File: examples/demo-react-native/package.json [hunk 18]
```diff
@@ -22,18 +22,30 @@
...
+    "@types/jest": "^29.2.1",
+    "@types/react": "^18.0.24",
+    "@types/react-test-renderer": "^18.0.0",
+    "babel-jest": "^29.2.1",
+    "eslint": "^8.19.0",
+    "jest": "^29.2.1",
+    "metro-react-native-babel-preset": "0.73.9",
+    "prettier": "^2.4.1",
+    "react-test-renderer": "18.2.0",
+    "typescript": "4.8.4"
```
Comment (asafkorem) [inline examples/demo-react-native/package.json:49]:
```suggestion
    "@tsconfig/react-native": "^2.0.2",
    "@types/jest": "^29.2.1",
    "@types/react": "^18.0.24",
    "jest": "^29.2.1",
    "typescript": "^4.8.4"
```

File: scripts/ci.android.sh [hunk 19]
```diff
@@ -18,27 +18,29 @@ pushd detox/test
...
-  run_f "npm run unit:android-release"
-  popd
-else
-  echo "Skipping Android unit tests (react-native version ${currentRnVersion} is not ≥66)"
+pushd detox/android
+run_f "npm run unit:android-release"
+popd
+
+NPM_ANDROID_FLAVOR="android"
+if [[ $currentRnVersion -lt 71 ]]; then
```
Comment (asafkorem) [inline scripts/ci.android.sh:26]: revert after fixing the previous comment

**Review by asafkorem — COMMENTED** [review]
GG us

**Review by d4vidi — APPROVED** [review]

**Comment by kevindice** [comment]
Thank you! 🎉

**Comment by RohovDmytro** [comment]
Wow! Great & epic.

When 0.72? 😜 I observe that going from 0.71 to 0.72 is doing not bad in the community. Me personally had a great migration experience in a moderately complex project.

**Comment by asafkorem** [comment]
> Wow! Great & epic.
> 
> When 0.72? 😜 I observe that going from 0.71 to 0.72 is doing not bad in the community. Me personally had a great migration experience in a moderately complex project.

Great question, we may try this soon and if the migration will be seamless we won't postpone it.. Anyhow, Detox might work with `0.72` now without changes. Feel free to try this and update
The new Detox version that officially supports React Native `0.71` is `20.9.1 `

**Comment by RohovDmytro** [comment]
> > Wow! Great & epic.
> > When 0.72? 😜 I observe that going from 0.71 to 0.72 is doing not bad in the community. Me personally had a great migration experience in a moderately complex project.
> 
> Great question, we may try this soon and if the migration will be seamless we won't postpone it.. Anyhow, Detox might work with `0.72` now without changes. Feel free to try this and update The new Detox version that officially supports React Native `0.71` is `20.9.1`

detox@20.9.1 + react-native@0.72.1 + Android = 👍, worked seamlessly from the first run
detox@20.9.1 + react-native@0.72.1 + iOS = investigating :)

**Comment by RohovDmytro** [comment]
Btw, you guys should know this, but react native 0.71 had a nasty, nasty, awful performance issue (people not screaming enough, but I do).

It had quite severe impact in a bigger project. My performance KPI's plummeted.

https://github.com/facebook/react-native/issues/35778

It was fixed in 0.72.0 ;) 🙂

**Comment by d4vidi** [comment]
> > Wow! Great & epic.
> > When 0.72? 😜 I observe that going from 0.71 to 0.72 is doing not bad in the community. Me personally had a great migration experience in a moderately complex project.
> 
> Great question, we may try this soon and if the migration will be seamless we won't postpone it.. Anyhow, Detox might work with `0.72` now without changes. Feel free to try this and update The new Detox version that officially supports React Native `0.71` is `20.9.1`

Correction: The official version is `20.10.0` 🚀

**Comment by d4vidi** [comment]
@RohovDmytro Thanks for this valuable input. We will definitely take that into account

**Comment by ruibinch** [comment]
@RohovDmytro Hello, just curious if you managed to get this new version of detox working seamlessly with RN 0.72.x on iOS? (will be performing the upgrade soon too)
